### PR TITLE
feat: add list pipeline template operations

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

Add operation list for `pipelineTemplate` meta and `pipelineTemplateVersions`

## Objective

Below changes have been made
- added list() function to `PipelineTemplateFactory`
- added list() function to `PipelineTemplateVersionsFactory`

## References

https://github.com/screwdriver-cd/screwdriver/issues/2135

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/templateMetaFactory.js [hunk 1]
```diff
@@ -23,18 +23,49 @@ class TemplateMetaFactory extends BaseFactory {
...
 
+    /**
+     * Create pipeline template meta
+     * @method list
+     * @param  {Object}   config                  Config object for template meta
+     * @param  {Object}   config.pipelineId       Identifier of the pipeline that is publishing the template
+     * @return {Promise}                          Resolve pipeline template meta
+     */
     create(config) {
         config.templateType = this._getTemplateType();
```
Comment (tkyi) [inline lib/templateMetaFactory.js]:
```suggestion
        config.templateType = this.getTemplateType();
```

File: lib/pipelineTemplateVersionFactory.js [hunk 2]
```diff
@@ -130,6 +130,46 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+            params.namespace = namespace;
+        }
+
+        const pipelineTemplateMeta = await templateMetaFactory.get(params);
+
+        if (!pipelineTemplateMeta) {
+            return [];
+        }
+
+        templateId = pipelineTemplateMeta.id;
```
Comment (tkyi) [inline lib/pipelineTemplateVersionFactory.js]: I don't love that we're potentially overwriting a param

File: lib/templateMetaFactory.js [hunk 3]
```diff
@@ -23,18 +23,49 @@ class TemplateMetaFactory extends BaseFactory {
...
+     * List all pipeline template meta
+     * @method list
+     * @param  {Object}   config                  Config object for template meta
+     * @param  {Object}   config.params           Parameters to filter on
+     * @return {Promise}                          Resolve pipeline template meta list
+     */
+    list(config) {
+        config.params.templateType = this.getTemplateType();
+
+        return super.list(config);
```
Comment (sagar1312) [inline lib/templateMetaFactory.js:71]: We should do something similar what is being done in `[TemplateTagFactory](https://github.com/screwdriver-cd/models/blob/master/lib/templateTagFactory.js#L149)`
```suggestion
        if (!config.params) {
            config.params = {};
        }
        config.params.templateType = this.getTemplateType();

        return super.list(config);
```

File: lib/pipelineTemplateFactory.js [hunk 4]
```diff
@@ -14,6 +14,21 @@ class PipelineTemplateFactory extends TemplateMetaFactory {
...
+     * @return {Promise}                          Resolve pipeline template meta
+     */
+    async list(config) {
+        if (config.params) {
+            config.params.templateType = this._getTemplateType();
+        }
+
+        return super.list(config);
+    }
+
```
Comment (sagar1312) [inline lib/pipelineTemplateFactory.js]: This is not needed. This should be abstracted into parent class `TemplateMetaFactory`
```suggestion
```

File: lib/pipelineTemplateVersionFactory.js (hunk 2 above)
Comment (sagar1312) [reply lib/pipelineTemplateVersionFactory.js]: Second that. We should make `name + namespace` and `templateId` mutually exclusive.

File: lib/pipelineTemplateVersionFactory.js [hunk 5]
```diff
@@ -130,6 +130,46 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+
+        return super.list({
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]: We may also have to pass other configs Ex: `config.paginate`

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: lib/pipelineTemplateVersionFactory.js (hunk 2 above)
Comment (pritamstyz4ever) [reply lib/pipelineTemplateVersionFactory.js]: I agree, let me fix that

File: lib/pipelineTemplateVersionFactory.js (hunk 5 above)
Comment (pritamstyz4ever) [reply lib/pipelineTemplateVersionFactory.js]: ok, so we do intend to support pagination and sorting? It's not in the current design doc.

**Review by tkyi — APPROVED** [review]

File: lib/pipelineTemplateVersionFactory.js (hunk 5 above)
Comment (sagar1312) [reply lib/pipelineTemplateVersionFactory.js]: It is not called out explicitly. The list api should allow filtering and sorting.

File: lib/pipelineTemplateVersionFactory.js [hunk 6]
```diff
@@ -130,6 +130,47 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
 
+    /**
+     * List all the versions of a pipeline template
+     * @method create
+     * @param  {Object}     config               Config object
+     * @param templateMetaFactory
+     * @param  {Datastore}  config.datastore     Object that will perform operations on the datastore
+     * @param  {String}     config.name          The template name
+     * @param  {String}     config.namespace     The template namespace
+     * @param  {String}     config.templateId    Identifier of the template
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]:
```suggestion
     * @param  {String}     config.params.templateId    Identifier of the template
```

File: lib/pipelineTemplateVersionFactory.js [hunk 7]
```diff
@@ -130,6 +130,47 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+            return [];
+        }
+
+        return super.list({
+            params: {
+                templateId: pipelineTemplateMeta.id
+            },
+            paginate: config.paginate
+        });
+    }
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]: The list API should be generic to allow filtering on any field. Accept filter criteria `config.params` and pass it with any modifications to the query.

```suggestion
        return super.list(config);
    }
```

File: lib/pipelineTemplateVersionFactory.js [hunk 8]
```diff
@@ -130,6 +130,47 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+        if (name && namespace) {
+            params.name = name;
+            params.namespace = namespace;
+        }
+
+        const pipelineTemplateMeta = await templateMetaFactory.get(params);
+
+        if (!pipelineTemplateMeta) {
+            return [];
+        }
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]: We could skip getting templateMeta when `name` and `namespace` are not specified.

```suggestion
        if (name && namespace) {
            const params = {};
        
            params.name = name;
            params.namespace = namespace;
            
            if (config.params.templateId) {
                params.id = config.params.templateId;
            }
            
            const pipelineTemplateMeta = await templateMetaFactory.get(params);
            
            if(pipelineTemplateMeta) {
               config.params.templateId = pipelineTemplateMeta.id;
            }
            else {
              return [];
            }
        }
```

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: lib/pipelineTemplateVersionFactory.js (hunk 8 above)
Comment (pritamstyz4ever) [reply lib/pipelineTemplateVersionFactory.js]: removing templateId, as per docs the API is to list the versions by name and namespace.

File: lib/pipelineTemplateVersionFactory.js [hunk 9]
```diff
@@ -130,6 +130,47 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+     * @param  {Object}     config.paginate             Pagination parameters
+     * @param  {Number}     config.paginate.count       Number of items per page
+     * @param  {Number}     config.paginate.page        Specific page of the set to return
+     * @return {Promise}                         List of PipelineTemplateVersionModel
+     */
+    async list(config, templateMetaFactory) {
+        const { name, namespace } = config.params;
+
+        if (!name || !namespace) {
+            return [];
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]: Returning empty list when only `name` or `namespace` is specified could be misleading
We should throw error indicating both `name` and `namespace` is mandatory.

File: lib/pipelineTemplateVersionFactory.js [hunk 10]
```diff
@@ -130,6 +130,47 @@ class PipelineTemplateVersionFactory extends BaseFactory {
...
+        }
+
+        return super.list({
+            params: {
+                templateId: pipelineTemplateMeta.id
+            },
+            paginate: config.paginate,
+            sort: config.sort
+        });
```
Comment (sagar1312) [inline lib/pipelineTemplateVersionFactory.js]: Plucking specific config makes the API rigid and not extensible.
List API should support [all the configs](https://github.com/screwdriver-cd/models/blob/master/lib/baseFactory.js#L110) supported by the base class implementation.
```suggestion
        if(!config.params) {
          config.params = {};
        }
        
        config.params.templateId = pipelineTemplateMeta.id;
        
        return super.list(config);
```

**Review by sagar1312 — DISMISSED** [review]

**Review by jithine — APPROVED** [review]

**Pull request merged**